# Mod 4 Project - Starter Notebook

This notebook has been provided to you so that you can make use of the following starter code to help with the trickier parts of preprocessing the Zillow dataset. 

The notebook contains a rough outline the general order you'll likely want to take in this project. You'll notice that most of the areas are left blank. This is so that it's more obvious exactly when you should make use of the starter code provided for preprocessing. 

**_NOTE:_** The number of empty cells are not meant to infer how much or how little code should be involved in any given step--we've just provided a few for your convenience. Add, delete, and change things around in this notebook as needed!

# Step 1: Load the Data/Filtering for Chosen Zipcodes

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

df= pd.read_csv('zillow_data.csv')

df.head()



# df.iloc[:,-1]

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-07  2017-08  2017-09  \
0         1  334200.0  335400.0  336500.0  ...  1005500  1007500  1007800   
1         2  235700.0  236900.0  236700.0  ...   308000   310000   312500   
2         3  210400.0  212200.0  212200.0  ...   321000   320600   320200   
3         4  498100.0  500900.0  503100.0  ...  1289800  1287700  1287400   
4         5   77300.0   77300.0   77300.0  ...   119100   119400   120000   

   2017-10  2017-11  2017-12  2018-01  2018-02  2018-03  2018-04  
0  1009600  1013300  1018700  1024400  1030700  1033800  1030600  
1   314100   315000   316600   318100   319600   321100   321800  
2   320400   320800   321200   321200   323000   326900   329900  
3  1291500  1296600  1299000  1302700  1306400  1308500  1307000  
4   120300   120300   120300   120300   120500   121000   121500  

[5 rows x 272 columns]

In [2]:
df['StdPrice']= df.std(skipna=True, axis=1)

In [5]:
df['AvgHomePrice']= df.mean(skipna=True, axis=1)

In [6]:
df['ROI']= (df.iloc[:,-3]-df.iloc[:,7])/df.iloc[:,7]

In [7]:
df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06  ...  2017-10  2017-11  2017-12  \
0         1  334200.0  335400.0  336500.0  ...  1009600  1013300  1018700   
1         2  235700.0  236900.0  236700.0  ...   314100   315000   316600   
2         3  210400.0  212200.0  212200.0  ...   320400   320800   321200   
3         4  498100.0  500900.0  503100.0  ...  1291500  1296600  1299000   
4         5   77300.0   77300.0   77300.0  ...   120300   120300   120300   

   2018-01  2018-02  2018-03  2018-04       StdPrice   AvgHomePrice       ROI  
0  1024400  1030700  1033800  1030600  203451.914731  734212.505259  2.083782  
1   318100   319600   321100   321800   37815.520226  217146.303049  0.365295  
2   321200   323000   326900   329900   43539.184162  259004.900313  0.567966  
3  1302700  1306400  1308500  1307000  249800.040104  961122.431376  1.623971  
4   120300   120500   121000   121500   19158.408568  101075.254307  0.571798  

[5 rows x 275 columns]

Split the data into different investment brackets:
1. 1M>
2. 750K<x<500K
3. 500K<

In [10]:
df_melted=pd.melt(df, id_vars=['RegionName', 'City', 'Metro','AvgHomePrice', 'ROI', 'StdPrice'], value_vars=df.iloc[:,7:-3], var_name='Date', value_name='Price')

In [12]:
current_price=df_melted[df_melted['Date']=='2018-04']

In [34]:
# df_melted_group= df_melted.groupby('RegionName').mean()
# df_melted[df_melted['City']=='Chicago']


In [33]:
# index_map_region= df[['RegionName','City', 'Metro']]
# pd.concat([index_map_region, df_melted_group], axis=1, join='outer', keys= 'RegionName')

In [22]:
top_10_Std= top_10_ROI= df[['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', 'SizeRank', 'AvgHomePrice', 'ROI', 'StdPrice']].sort_values('StdPrice', ascending= True).head(25)

In [23]:
top_10_Std

RegionID  RegionName              City State  \
13091     74754       38317          Bruceton    TN   
13895     75432       39813         Arlington    GA   
7249      63502       14301     Niagara Falls    NY   
13181     74734       38255            Sharon    TN   
5389      90049       73521             Altus    OK   
14266     74774       38342       Hollow Rock    TN   
14088     86209       63837          Clarkton    MO   
9045      68717       25702        Huntington    WV   
14611     82427       55753          Keewatin    MN   
10817     70310       29112             North    SC   
14385     76354       42410        Earlington    KY   
8753      76900       43935     Martins Ferry    OH   
3711      63941       15210        Pittsburgh    PA   
6181      63506       14305     Niagara Falls    NY   
14310     85616       62665         Meredosia    IL   
4517      76890       43920    East Liverpool    OH   
12435     70307       29107            Neeses    SC   
14216     74737       38258         Trezevant    TN   
6559      77821       45804              Lima    OH   
9376      90478       74631         Blackwell    OK   
11071     63857       15045         Glassport    PA   
12028     86031       63382          Vandalia    MO   
14584     85326       62094              Witt    IL   
7070      89009       71295         Winnsboro    LA   
8940      81732       54495  Wisconsin Rapids    WI   

                             Metro  CountyName  SizeRank  AvgHomePrice  \
13091                          NaN     Carroll     13092  41844.091252   
13895                          NaN     Calhoun     13896  34857.148198   
7249                       Buffalo     Niagara      7250  33504.286240   
13181                       Martin     Weakley     13182  43880.803178   
5389                         Altus     Jackson      5390  48864.316388   
14266                          NaN     Carroll     14267  41400.129565   
14088                      Kennett     Dunklin     14089  44647.804658   
9045                    Huntington      Cabell      9046  45270.602060   
14611                          NaN      Itasca     14612  41156.171403   
10817                   Orangeburg  Orangeburg     10818  44153.300777   
14385                 Madisonville     Hopkins     14386  22274.156380   
8753                      Wheeling     Belmont      8754  49760.015320   
3711                    Pittsburgh   Allegheny      3712  37755.379269   
6181                       Buffalo     Niagara      6182  49960.453564   
14310                 Jacksonville      Morgan     14311  38622.909718   
4517                         Salem  Columbiana      4518  50232.829975   
12435                   Orangeburg  Orangeburg     12436  42224.786741   
14216                          NaN     Carroll     14217  46582.278067   
6559                          Lima       Allen      6560  29299.502076   
9376                    Ponca City         Kay      9377  30847.168709   
11071                   Pittsburgh   Allegheny     11072  40669.241470   
12028                       Mexico     Audrain     12029  56346.455401   
14584                          NaN  Montgomery     14585  24848.969692   
7070                           NaN    Franklin      7071  51633.388120   
8940   Wisconsin Rapids-Marshfield        Wood      8941  73732.524397   

            ROI     StdPrice  
13091  0.340599  4097.546840  
13895  0.420339  4131.865394  
7249   0.246575  4299.998433  
13181  0.216749  4365.054892  
5389   0.133475  4441.108417  
14266  0.354286  4451.852879  
14088       NaN  4618.135908  
9045        NaN  4790.125411  
14611  0.204188  4818.107423  
10817  0.121429  4897.908961  
14385  0.540881  4898.066170  
8753        NaN  5033.328588  
3711   0.476048  5134.023468  
6181   0.353306  5269.008791  
14310  0.411243  5270.714274  
4517   0.504000  5303.263204  
12435  0.101852  5317.633347  
14216  0.295285  5320.800139  
6559  -0.004032  5381.058519  
9376   0.417004  5402.382849  


In [27]:
top_10_ROI= df[['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', 'SizeRank', 'AvgHomePrice', 'ROI','StdPrice']].sort_values('ROI', ascending= False).head(25)

In [28]:
top_10_ROI

RegionID  RegionName           City State  \
117       62022       11211       New York    NY   
1155      62033       11222       New York    NY   
475       62027       11216       New York    NY   
191       60639        7302    Jersey City    NJ   
106       62026       11215       New York    NY   
13864     62320       11975      Wainscott    NY   
11728     62281       11930     Amagansett    NY   
14217     93399       80449        Hartsel    CO   
1790      96127       90291    Los Angeles    CA   
258       66125       20001     Washington    DC   
12614     62283       11932  Bridgehampton    NY   
11172     62321       11976     Water Mill    NY   
6048      83143       57110    Sioux Falls    SD   
3615      65792       19125   Philadelphia    PA   
7500      58634        2120         Boston    MA   
13830     66102       19951       Harbeson    DE   
70        62032       11221       New York    NY   
6650      71037       30317        Atlanta    GA   
3926      92615       78702         Austin    TX   
4678      70448       29403     Charleston    SC   
3339      97696       94306      Palo Alto    CA   
76        97565       94110  San Francisco    CA   
9296      70205       28762       Old Fort    NC   
7007      62310       11963         Noyack    NY   
3669      58633        2119         Boston    MA   

                                Metro            CountyName  SizeRank  \
117                          New York                 Kings       118   
1155                         New York                 Kings      1156   
475                          New York                 Kings       476   
191                          New York                Hudson       192   
106                          New York                 Kings       107   
13864                        New York               Suffolk     13865   
11728                        New York               Suffolk     11729   
14217                          Denver                  Park     14218   
1790   Los Angeles-Long Beach-Anaheim           Los Angeles      1791   
258                        Washington  District of Columbia       259   
12614                        New York               Suffolk     12615   
11172                        New York               Suffolk     11173   
6048                      Sioux Falls             Minnehaha      6049   
3615                     Philadelphia          Philadelphia      3616   
7500                           Boston               Suffolk      7501   
13830                       Salisbury                Sussex     13831   
70                           New York                 Kings        71   
6650                          Atlanta                Fulton      6651   
3926                           Austin                Travis      3927   
4678                       Charleston            Charleston      4679   
3339                         San Jose           Santa Clara      3340   
76                      San Francisco         San Francisco        77   
9296                           Marion              McDowell      9297   
7007                         New York               Suffolk      7008   
3669                           Boston               Suffolk      3670   

       AvgHomePrice        ROI       StdPrice  
117    5.880672e+05  11.189940  397436.751042  
1155   6.324746e+05  10.535523  448664.362986  
475    5.653764e+05   9.942505  387033.827390  
191    5.891678e+05   9.403061  321308.640603  
106    1.061516e+06   8.941958  609481.339589  
13864  1.554144e+06   8.922175  806598.720705  
11728  1.444255e+06   8.564860  743531.176567  
14217  1.445385e+05   7.923333   58087.412604  
1790   9.164767e+05   7.778489  518516.867575  
258    3.803046e+05   7.622826  205447.162329  
12614  1.218150e+06   7.276181  570189.388592  
11172  1.779875e+06   7.103500  798795.794725  
6048   1.611715e+05   7.102990   52136.147004  
3615   1.070393e+05   7.046012   64831.506488  
7500   3.213665e+05   7.029545  147

In [24]:
top_10_AvgHomePrice= df[['RegionID', 'RegionName', 'City', 'State', 'Metro', 'CountyName', 'SizeRank', 'AvgHomePrice', 'ROI', 'StdPrice']].sort_values('AvgHomePrice', ascending= False).head(25)

In [26]:
top_10_AvgHomePrice

RegionID  RegionName              City State  \
272       61635       10021          New York    NY   
20        61625       10011          New York    NY   
508       61628       10014          New York    NY   
21        61703       10128          New York    NY   
10237     97518       94027          Atherton    CA   
7596      93816       81611             Aspen    CO   
4816      96086       90210     Beverly Hills    CA   
4723      72636       33480        Palm Beach    FL   
2026      97577       94123     San Francisco    CA   
742       97569       94115     San Francisco    CA   
9         97564       94109     San Francisco    CA   
14622     71578       31561        Sea Island    GA   
12098     93818       81615  Snowmass Village    CO   
7380      96149       90402      Santa Monica    CA   
8048      96621       92067   Rancho Santa Fe    CA   
985       96001       90020       Los Angeles    CA   
10008     97519       94028    Portola Valley    CA   
5738      97691       94301         Palo Alto    CA   
13900     60722        7620            Alpine    NJ   
8134      96975       92657     Newport Coast    CA   
5843      97513       94022         Los Altos    CA   
13144     97895       94957              Ross    CA   
7827      97130       93108         Montecito    CA   
1534      62028       11217          New York    NY   
8256      97695       94305          Stanford    CA   

                                Metro     CountyName  SizeRank  AvgHomePrice  \
272                          New York       New York       273  1.258371e+07   
20                           New York       New York        21  7.584343e+06   
508                          New York       New York       509  6.694651e+06   
21                           New York       New York        22  5.015868e+06   
10237                   San Francisco      San Mateo     10238  3.441001e+06   
7596                 Glenwood Springs         Pitkin      7597  3.104532e+06   
4816   Los Angeles-Long Beach-Anaheim    Los Angeles      4817  2.753846e+06   
4723            Miami-Fort Lauderdale     Palm Beach      4724  2.599595e+06   
2026                    San Francisco  San Francisco      2027  2.596361e+06   
742                     San Francisco  San Francisco       743  2.367470e+06   
9                       San Francisco  San Francisco        10  2.363896e+06   
14622                       Brunswick          Glynn     14623  2.345880e+06   
12098                Glenwood Springs         Pitkin     12099  2.268953e+06   
7380   Los Angeles-Long Beach-Anaheim    Los Angeles      7381  2.262241e+06   
8048                        San Diego      San Diego      8049  2.140756e+06   
985    Los Angeles-Long Beach-Anaheim    Los Angeles       986  2.121550e+06   
10008                   San Francisco      San Mateo     10009  2.103085e+06   
5738                         San Jose    Santa Clara      5739  2.057809e+06   
13900                        New York         Bergen     13901  2.051881e+06   
8134   Los Angeles-Long Beach-Anaheim         Orange      8135  2.042846e+06   
5843                         San Jose    Santa Clara      5844  2.039875e+06   
13144                   San Francisco          Marin     13145  2.016754e+06   
7827        Santa Maria-Santa Barbara  Santa Barbara      7828  1.965518e+06   
1534                         New York          Kings      1535  1.937322e+06   
8256                         San Jose    Santa Clara      8257  1.893734e+06   

            ROI      StdPrice  
272         NaN  3.727137e+06  
20          NaN  2.825544e+06  
508         NaN  2.239233e+06  
21     1.009030  1.556229e+06  
10237  4.763653  1.338385e+06  
7596   2.303028  9.480534e+05  
4816   4.866358  1.249521e+06  
4723   3.889608  1.038093e+06  
2026   4.514427  1.018333e+06  
742    4.239924  9.140307e+05  
9      3.978460  8.529518e+05  
14622       NaN  4.394686e+05  
12098  0.991222  6.134466e+05  
7380   3.748864  9.075405e+05  
8048   1.896872

Choose 11211 (Highest ROI and one of 4 King County NY zips in top 5 of ROI), 11930 (7th highest ROI, NY Suffolk County, Beach property), 80449 (Hartsel CO, 8th highest ROI), 20001 Washington DC (10th highest ROI), 94360 (Palo Alto, CA 21st ROI 18th on Avg Home Price)

Diversity in the locations of the houses, we have one west coast zip followed with one midwest and then three east coast cities. Statyed in the major cities due to it being where the greatest ROI has been seen over the past 25 years. Also, there is less risk in the cities due to the trend of millenials along with gen z migratiing to cities.  People will always need places to live and the properties will only increase in property value due to growing demand. m

# Step 2: Data Preprocessing

In [17]:
def get_datetimes(df):
    return pd.to_datetime(df.columns.values[1:], format='%Y-%m')

# Step 3: EDA and Visualization

In [ ]:
font = {'family' : 'normal',
        'weight' : 'bold',
        'size'   : 22}

matplotlib.rc('font', **font)

# Step 4: Reshape from Wide to Long Format

In [ ]:
def melt_data(df):
    melted = pd.melt(df, id_vars=['RegionName', 'City', 'State', 'Metro', 'CountyName'], var_name='time')
    melted['time'] = pd.to_datetime(melted['time'], infer_datetime_format=True)
    melted = melted.dropna(subset=['value'])
    return melted.groupby('time').aggregate({'value':'mean'})

# Step 5: ARIMA Modeling

# Step 6: Interpreting Results